rust/kernel/thread.rs

```diff
@@ -0,0 +1,307 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! A kernel thread (kthread).
+//!
+//! This module allows Rust code to create/wakeup/stop a kernel thread.
+
+use crate::c_types;
+use crate::error::{from_kernel_err_ptr, Error, Result};
+use crate::str::CStr;
+use crate::task::{Task, TaskRef};
+use crate::{bindings, c_str};
+
+use alloc::boxed::Box;
+use core::ops::FnOnce;
+
+/// Raw kernel threads.
+///
+/// The threads are created via C-style functions and machine-word sized
+/// arguments.
+///
+/// # Safety
+///
+/// The creation of [`RawThread`] is unsafe because if a caller provides
+/// an incorrect thread function or an incorrect thread argument, the
+/// new thread will corrupt memory or do other unsafe behaviors.
+pub struct RawThread {
+    task: Task,
+}
+
+impl RawThread {
+    /// Creates a new thread using a C-style function pointer.
+    ///
+    /// # Safety
+    ///
+    /// This function actually doesn't dereference `arg` or call `f`, so even if
+    /// the users pass incorrect parameters this function won't run into
+    /// trouble. But if the users provide incorrect `arg` or `f` the new
+    /// thread will corrupt memory or do other unsafe behaviors, so
+    /// make it `unsafe`.
+    ///
+    /// The safety requirements of calling this function are:
+    ///
+    /// - Make sure `arg` is a proper pointer that points to a valid memory
+    ///   location when the new thread begins to run.
+    ///
+    /// - Make sure `f` is a proper function pointer and `f` handles `arg`
+    ///   correctly.
+    ///
+    /// # Context
+    ///
+    /// This function might sleep in `kthread_create_on_node` due to the memory
+    /// allocation and waiting for the completion, therefore do not call this
+    /// in atomic contexts (i.e. preemption-off contexts).
+    pub unsafe fn try_new(
+        name: &CStr,
+        f: unsafe extern "C" fn(*mut c_types::c_void) -> c_types::c_int,
+        arg: *mut c_types::c_void,
+    ) -> Result<Self> {
+        // SAFETY: `kthread_create_on_node` will copy the content of `name`,
+        // so we don't need to make the `name` live longer.
+        let task_ptr = from_kernel_err_ptr(unsafe {
+            bindings::kthread_create_on_node(
+                Some(f),
+                arg,
+                bindings::NUMA_NO_NODE,
+                c_str!("%s").as_ptr() as _,
```

Comment:
Agreed, but that requires the API implemented as a macro, right?

Reply:
> Agreed, but that requires the API implemented as a macro, right?

Maybe not. Instead of taking a `&CStr` as argument, you'd take a `core::fmt::Arguments`. Callers would have to use the `format_args!` macro to create the name, but they already had to use `c_str!` anyway; if `format_args!` is too long, perhaps we could define a new one called `fmt!` that just calls `format_args!`.

And the implementation would pass `c_str!("%pA")` as the format, with a pointer to the `Arguments` arg as the first (and only) format argument.

Perhaps you could do it in this PR if you're so inclined. But I'm fine with deferring it too.